gwu-libraries/scholarspace-hyrax
# T488 saml sso

In this PR:
 - Adds sign-on route for authenticating users via SAML, and a single-logout (SLO) process for SAML-authenticated users.
 - Removes registration link/route to prevent non-SAML users from registering.
 - Prevents authenticated, non-privileged users from accessing the dashboard.
 - Includes Rake task to update existing GW users to SAML accounts. 

You won't be able to test the SSO process on your dev instance. But with the addition of a `Sign in with SAML` button on the login page (which will take you to `example.com`), all other functionality should be intact. 

### Configuration steps:
1. Make sure your `.env` file defines an `NGINX_CERT_DIR` and `NGINX_KEY_DIR`, and that these point to directories on your server, even if you're not using HTTPS. (These directories will be used for the SAML cert/key as well.)
2. Make sure that `PERM_URL_BASE` is defined in the `.env` file, and that it ends with a forward slash. 
3. Add the variables from the `SAML` section in `example.env` to your `.env` file. You can leave the values as they are, however. (The Devise initializer will automatically create empty cert/key files if these are not found in the cert and key directories. However, the directories themselves must exist.)
4. Run `RAILS_ENV=production rails db:migrate` in the app container.
5. To test an authenticated user's access, you can create a user manually in the console:
```
  User.create(email: "testuser@example.com", password: "password")
```
6. If you haven't done so, make sure your repository contains private, public, and GW-only works.
 
### Testing
1. Sign in as an `admin` or `content-admin` user. Confirm that the dashboard is visible and functional.
2. Confirm that you can logout as expected.
3. Sign in as your newly created user (no elevated privileges).  
   a. Confirm that you are taken not to the dashboard but to the homepage.
   b. Confirm that this user *can* view GW-only works as well as public works.
   c. Confirm that this user *cannot* access any dashboard pages, such as `/dashboard`, `/dashboard/my/works`, etc. All such routes should redirect to the homepage.
   d. Confirm that search and browse functionality works as expected for this user.
   e. Log out as this user.
4. On the `Staff sign in` page, confirm that there is no link for registering a new user.
5. Access the app container and run the following command:
    ```
    RAILS_ENV=production rake gwss:update_users_sso
    ```
    You should see output indicating that your non-elevated test user has been deleted (and that any users with a GW email address have been updated). 
7. Try to log in with the test user again. You should see a message to the effect that the user does not exist.
8. Confirm that all spec tests pass in the app container.
9. Behavior should be the same in development and production, since your instance is not actually configured for SAML authentication.

**Note**: This PR includes most of the changes from Max's outstanding PR.
